jobs: report if any jobs in a txn failed.

spaskob opened the pull request:
Jobs queued in a txn can fail. Report any failures when commit phase is over. Also eagerly schedule the jobs.

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/43321)
<!-- Reviewable:end -->

thoszhang reviewed (commented):
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43321)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @spaskob)

___
*[pkg/jobs/jobs_test.go, line 1690 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43321#-Lxwuvkq8pHJglZVoCNs:-Lxwuvkq8pHJglZVoCNt:b-u876fa) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc7162944410a30f608572be980938319dc6ff0/pkg/jobs/jobs_test.go#L1690)):*
> ```Go
> 			OnResume: func() error {
> 				// Inc instead of just storing 1 to count how many jobs ran.
> 				// atomic.AddInt32(&hasRun, 1)
> ```

I'd just remove the lines being commented out.
___
*[pkg/jobs/registry.go, line 295 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43321#-Lxx2-Oe4So0jesuHPhl:-Lxx2-Oe4So0jesuHPhm:b-o9azv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc7162944410a30f608572be980938319dc6ff0/pkg/jobs/registry.go#L295)):*
> ```Go
> 		j, err := r.LoadJob(ctx, id)
> 		if err != nil {
> 			errMsg += fmt.Sprintf("Job %d could not be loaded. The job may not have succeeded", jobs[i])
> ```

I don't think we should be trying to combine all the job errors into a single error string. If there's only one error (which is probably the most common case, since there's usually only one job), then we could just use `errors.Wrap()`, which preserves the original error.

Otherwise, if there are multiple errors, I think we should either (1) just wrap and return the first error, (2) use the `errors` API to combine the errors (possibly with `errors.CombineErrors()`? https://github.com/cockroachdb/errors#available-wrapper-constructors), or (3) just return an error indicating that multiple jobs failed and they need to manually inspect the jobs table.


<!-- Sent from Reviewable.io -->

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43321)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @lucy-zhang)

___
*[pkg/jobs/jobs_test.go, line 1690 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43321#-Lxwuvkq8pHJglZVoCNs:-Ly4tkGeBgYx5APobVlV:bb8iijr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc7162944410a30f608572be980938319dc6ff0/pkg/jobs/jobs_test.go#L1690)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I'd just remove the lines being commented out.
</blockquote></details>

thanks, I've left this when debugging the testF
___
*[pkg/jobs/registry.go, line 295 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43321#-Lxx2-Oe4So0jesuHPhl:-Ly54LBwDEBYFt0wn_P6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc7162944410a30f608572be980938319dc6ff0/pkg/jobs/registry.go#L295)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I don't think we should be trying to combine all the job errors into a single error string. If there's only one error (which is probably the most common case, since there's usually only one job), then we could just use `errors.Wrap()`, which preserves the original error.

Otherwise, if there are multiple errors, I think we should either (1) just wrap and return the first error, (2) use the `errors` API to combine the errors (possibly with `errors.CombineErrors()`? https://github.com/cockroachdb/errors#available-wrapper-constructors), or (3) just return an error indicating that multiple jobs failed and they need to manually inspect the jobs table.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spaskob committed:
jobs: report if any jobs in a txn failed

Jobs queued in a txn can fail. Report any failures when commit phase is over.

Release note: none.

thoszhang reviewed (approved):
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43321)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)



<!-- Sent from Reviewable.io -->

spaskob commented:
bors r+
